Sources/NIOHPACK/StringRing.swift

```diff
@@ -0,0 +1,355 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+enum RingBufferError {
+    /// Error type thrown when a write would overrun the bounds of a
+    /// CircularByteBuffer.
+    struct BufferOverrun : Error, Equatable {
+        /// The amount by which the write would overrun.
+        let amount: Int
+    }
+}
+
+struct StringRing {
+    @_versioned private(set) var _storage: ByteBuffer
+    @_versioned private(set) var _ringHead = 0
+    @_versioned private(set) var _ringTail = 0
+    @_versioned private(set) var _readableBytes = 0
+    
+    // private but tests
+    @_inlineable @_versioned
```

Comment:
why do we have `@_inlineable` and `@_versioned` here? the `struct StringRing` is not `public`, nor do they use generics or closures so I don't understand why they would need to be inlined across a module boundary.

Reply:
They're an artifact of an earlier commit where I tried to have this be a general-purpose analogue to `ByteBuffer`. I should've removed them when I removed `public`.